torch/optim/adam.py

```diff
@@ -373,23 +378,41 @@ def _single_tensor_adam(params: List[Tensor],
                 (param.is_cuda and step_t.is_cuda) or (param.is_xla and step_t.is_xla)
             ), "If capturable=True, params and state_steps must be CUDA or XLA tensors."
 
-        # update step
-        step_t += 1
+        if not decoupled_weight_decay:
+            # update step
+            step_t += 1
 
-        if weight_decay != 0:
-            grad = grad.add(param, alpha=weight_decay)
+            if weight_decay != 0:
+                grad = grad.add(param, alpha=weight_decay)
 
-        if torch.is_complex(param):
-            grad = torch.view_as_real(grad)
-            exp_avg = torch.view_as_real(exp_avg)
-            exp_avg_sq = torch.view_as_real(exp_avg_sq)
+            if torch.is_complex(param):
```

Comment:
There's some duplication here as I figured the order of the update step and torch.is_complex() checks matter, can of course factor it out if it doesn't.

Reply:
To my understanding, the order shouldn't, so go ahead with the refactor!